In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score,confusion_matrix,recall_score,precision_score,f1_score
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.metrics import accuracy_score, classification_report

In [2]:
df=pd.read_csv('Crop_recommendation.csv')
df

N   P   K  temperature   humidity        ph    rainfall   label
0      90  42  43    20.879744  82.002744  6.502985  202.935536    rice
1      85  58  41    21.770462  80.319644  7.038096  226.655537    rice
2      60  55  44    23.004459  82.320763  7.840207  263.964248    rice
3      74  35  40    26.491096  80.158363  6.980401  242.864034    rice
4      78  42  42    20.130175  81.604873  7.628473  262.717340    rice
...   ...  ..  ..          ...        ...       ...         ...     ...
2195  107  34  32    26.774637  66.413269  6.780064  177.774507  coffee
2196   99  15  27    27.417112  56.636362  6.086922  127.924610  coffee
2197  118  33  30    24.131797  67.225123  6.362608  173.322839  coffee
2198  117  32  34    26.272418  52.127394  6.758793  127.175293  coffee
2199  104  18  30    23.603016  60.396475  6.779833  140.937041  coffee

[2200 rows x 8 columns]

In [3]:
df['label'].value_counts()

rice           100
maize          100
jute           100
cotton         100
coconut        100
papaya         100
orange         100
apple          100
muskmelon      100
watermelon     100
grapes         100
mango          100
banana         100
pomegranate    100
lentil         100
blackgram      100
mungbean       100
mothbeans      100
pigeonpeas     100
kidneybeans    100
chickpea       100
coffee         100
Name: label, dtype: int64

In [4]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()

In [5]:
from sklearn.preprocessing import LabelEncoder

# Create and fit the label encoder
label_encoder = LabelEncoder()
label_encoder.fit(df['label'])

# Get the mapping of each value to its encoded label
encoded_labels = label_encoder.transform(df['label'])
label_mapping = {label: original for label, original in zip(encoded_labels, df['label'])}

# Print the mapping
for encoded, original in label_mapping.items():
    print(f"{original}: {encoded}")


rice: 20
maize: 11
chickpea: 3
kidneybeans: 9
pigeonpeas: 18
mothbeans: 13
mungbean: 14
blackgram: 2
lentil: 10
pomegranate: 19
banana: 1
mango: 12
grapes: 7
watermelon: 21
muskmelon: 15
apple: 0
orange: 16
papaya: 17
coconut: 4
cotton: 6
jute: 8
coffee: 5


In [6]:
df['label']=label_encoder.fit_transform(df['label'])

encoded value of crop :


rice: 20

maize: 11

chickpea: 3

kidneybeans: 9

pigeonpeas: 18

mothbeans: 13

mungbean: 14

blackgram: 2

lentil: 10

pomegranate: 19

banana: 1

mango: 12

grapes: 7

watermelon: 21

muskmelon: 15

apple: 0

orange: 16

papaya: 17

coconut: 4

cotton: 6

jute: 8

coffee: 5

In [7]:
a=df
a=a.drop(columns='label')

In [8]:
X=a
y=df['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Logistic Regression

In [9]:
from sklearn.linear_model import LogisticRegression

# Initialize and train Logistic Regression
log_reg = LogisticRegression(max_iter=1000, random_state=42)
log_reg.fit(X_train, y_train)

# Make predictions
y_pred_log_reg = log_reg.predict(X_test)

# Evaluate
print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_log_reg)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_log_reg, target_names=label_encoder.classes_))


Logistic Regression Accuracy: 95.0
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.85      0.85      0.85        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       0.94      1.00      0.97        17
      cotton       0.80      0.94      0.86        17
      grapes       1.00      1.00      1.00        14
        jute       0.87      0.87      0.87        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       0.77      0.91      0.83        11
       maize       0.94      0.81      0.87        21
       mango       1.00      1.00      1.00        19
   mothbeans       0.91      0.88      0.89        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange       1.0

C:\Users\ATHARVA\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# Decision Tree

In [10]:
from sklearn.tree import DecisionTreeClassifier

# Initialize and train Decision Tree
tree_model = DecisionTreeClassifier(random_state=42)
tree_model.fit(X_train, y_train)

# Make predictions
y_pred_tree = tree_model.predict(X_test)

# Evaluate
print("Decision Tree Accuracy:", accuracy_score(y_test, y_pred_tree)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_tree, target_names=label_encoder.classes_))


Decision Tree Accuracy: 98.63636363636363
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.95      1.00      0.98        20
    chickpea       1.00      1.00      1.00        26
     coconut       0.96      1.00      0.98        27
      coffee       1.00      1.00      1.00        17
      cotton       1.00      1.00      1.00        17
      grapes       1.00      1.00      1.00        14
        jute       0.92      0.96      0.94        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       0.92      1.00      0.96        11
       maize       1.00      0.95      0.98        21
       mango       1.00      1.00      1.00        19
   mothbeans       1.00      0.92      0.96        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange   

# Random Forest

In [11]:
from sklearn.ensemble import RandomForestClassifier

# Initialize and train Random Forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions
y_pred_rf = rf_model.predict(X_test)

# Evaluate
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_rf, target_names=label_encoder.classes_))


Random Forest Accuracy: 99.31818181818181
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       1.00      1.00      1.00        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      1.00      1.00        17
      cotton       1.00      1.00      1.00        17
      grapes       1.00      1.00      1.00        14
        jute       0.92      1.00      0.96        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       0.92      1.00      0.96        11
       maize       1.00      1.00      1.00        21
       mango       1.00      1.00      1.00        19
   mothbeans       1.00      0.96      0.98        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange   

# Support Vector Machine (SVM)

In [12]:
from sklearn.svm import SVC

# Initialize and train SVM
svm_model = SVC(random_state=42)
svm_model.fit(X_train, y_train)

# Make predictions
y_pred_svm = svm_model.predict(X_test)

# Evaluate
print("SVM Accuracy:", accuracy_score(y_test, y_pred_svm)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_svm, target_names=label_encoder.classes_))


SVM Accuracy: 96.13636363636363
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.91      1.00      0.95        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      0.94      0.97        17
      cotton       0.85      1.00      0.92        17
      grapes       1.00      1.00      1.00        14
        jute       0.77      1.00      0.87        23
 kidneybeans       0.95      1.00      0.98        20
      lentil       0.79      1.00      0.88        11
       maize       1.00      0.86      0.92        21
       mango       1.00      1.00      1.00        19
   mothbeans       1.00      0.88      0.93        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange       1.00  

# K-Nearest Neighbors (KNN)

In [13]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize and train KNN
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

# Make predictions
y_pred_knn = knn_model.predict(X_test)

# Evaluate
print("KNN Accuracy:", accuracy_score(y_test, y_pred_knn)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_knn, target_names=label_encoder.classes_))


KNN Accuracy: 97.04545454545455
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.95      1.00      0.98        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      0.94      0.97        17
      cotton       0.94      1.00      0.97        17
      grapes       1.00      1.00      1.00        14
        jute       0.76      0.96      0.85        23
 kidneybeans       0.95      1.00      0.98        20
      lentil       0.85      1.00      0.92        11
       maize       1.00      0.95      0.98        21
       mango       1.00      1.00      1.00        19
   mothbeans       1.00      0.92      0.96        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange       1.00  

# Naive Bayes

In [14]:
from sklearn.naive_bayes import GaussianNB

# Initialize and train Naive Bayes
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)

# Make predictions
y_pred_nb = nb_model.predict(X_test)

# Evaluate
print("Naive Bayes Accuracy:", accuracy_score(y_test, y_pred_nb)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_nb, target_names=label_encoder.classes_))


Naive Bayes Accuracy: 99.54545454545455
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       1.00      1.00      1.00        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      1.00      1.00        17
      cotton       1.00      1.00      1.00        17
      grapes       1.00      1.00      1.00        14
        jute       0.92      1.00      0.96        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       1.00      1.00      1.00        11
       maize       1.00      1.00      1.00        21
       mango       1.00      1.00      1.00        19
   mothbeans       1.00      1.00      1.00        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange     

# Gradient Boosting (XGBoost)

In [15]:
from xgboost import XGBClassifier

# Initialize and train XGBoost
xgb_model = XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', random_state=42)
xgb_model.fit(X_train, y_train)

# Make predictions
y_pred_xgb = xgb_model.predict(X_test)

# Evaluate
print("XGBoost Accuracy:", accuracy_score(y_test, y_pred_xgb)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_xgb, target_names=label_encoder.classes_))


XGBoost Accuracy: 98.63636363636363
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.95      1.00      0.98        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       1.00      1.00      1.00        17
      cotton       0.94      1.00      0.97        17
      grapes       1.00      1.00      1.00        14
        jute       0.96      0.96      0.96        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       0.92      1.00      0.96        11
       maize       1.00      0.95      0.98        21
       mango       0.95      1.00      0.97        19
   mothbeans       1.00      0.96      0.98        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange       1.

# Neural Network (MLPClassifier)

In [16]:
from sklearn.neural_network import MLPClassifier

# Initialize and train Neural Network
mlp_model = MLPClassifier(max_iter=1000, random_state=42)
mlp_model.fit(X_train, y_train)

# Make predictions
y_pred_mlp = mlp_model.predict(X_test)

# Evaluate
print("Neural Network Accuracy:", accuracy_score(y_test, y_pred_mlp)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_mlp, target_names=label_encoder.classes_))


Neural Network Accuracy: 95.45454545454545
Classification Report:
               precision    recall  f1-score   support

       apple       1.00      1.00      1.00        23
      banana       1.00      1.00      1.00        21
   blackgram       0.84      0.80      0.82        20
    chickpea       1.00      1.00      1.00        26
     coconut       1.00      1.00      1.00        27
      coffee       0.94      1.00      0.97        17
      cotton       0.81      1.00      0.89        17
      grapes       1.00      1.00      1.00        14
        jute       0.87      0.87      0.87        23
 kidneybeans       1.00      1.00      1.00        20
      lentil       0.83      0.91      0.87        11
       maize       1.00      0.76      0.86        21
       mango       1.00      1.00      1.00        19
   mothbeans       0.85      0.92      0.88        24
    mungbean       1.00      1.00      1.00        19
   muskmelon       1.00      1.00      1.00        17
      orange  

# Deep Learning (Keras)

In [17]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

# Convert target to categorical (one-hot encoding)
y_train_categorical = to_categorical(y_train)
y_test_categorical = to_categorical(y_test)

# Initialize and train Deep Learning model
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(len(label_encoder.classes_), activation='softmax'))

model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X_train, y_train_categorical, epochs=50, batch_size=10, verbose=1, validation_split=0.2)

# Make predictions
y_pred_dl = model.predict(X_test)
y_pred_dl_classes = y_pred_dl.argmax(axis=1)

# Evaluate
print("Deep Learning Accuracy:", accuracy_score(y_test, y_pred_dl_classes)*100)
print("Classification Report:\n", classification_report(y_test, y_pred_dl_classes, target_names=label_encoder.classes_))

Epoch 1/50
141/141 [==============================] - 1s 4ms/step - loss: 4.4881 - accuracy: 0.4688 - val_loss: 0.6351 - val_accuracy: 0.8267
Epoch 2/50
141/141 [==============================] - 0s 3ms/step - loss: 0.4500 - accuracy: 0.8523 - val_loss: 0.3171 - val_accuracy: 0.8864
Epoch 3/50
141/141 [==============================] - 0s 3ms/step - loss: 0.4020 - accuracy: 0.8629 - val_loss: 0.3474 - val_accuracy: 0.8665
Epoch 4/50
141/141 [==============================] - 0s 3ms/step - loss: 0.2767 - accuracy: 0.9041 - val_loss: 0.1639 - val_accuracy: 0.9517
Epoch 5/50
141/141 [==============================] - 0s 3ms/step - loss: 0.2457 - accuracy: 0.9212 - val_loss: 0.1544 - val_accuracy: 0.9489
Epoch 6/50
141/141 [==============================] - 0s 3ms/step - loss: 0.2143 - accuracy: 0.9190 - val_loss: 0.3045 - val_accuracy: 0.9034
Epoch 7/50
141/141 [==============================] - 0s 3ms/step - loss: 0.2253 - accuracy: 0.9169 - val_loss: 0.1776 - val_accuracy: 0.9290
Epoch 

Logistic Regression ==>95.0

Decision Tree ==>98.63636363636363

Random Forest==>99.31818181818181

Support Vector Machine ==> 96.13636363636363

K-Nearest Neighbors==> 97.04545454545455

Naive Bayes ==> 99.54545454545455

Gradient Boosting ==> 98.63636363636363

Neural Network==>95.45454545454545

Deep Learning==>92.04545454545455